### PR TITLE
fix: Accesscontrolled layout

### PR DESCRIPTION
**Pull request by caalador**
Make Flow (a)Layout linking dynamic.
Add access checking for route Layout.

Part of #20097

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/auth/AnnotatedViewAccessChecker.java [hunk 1]
```diff
@@ -60,6 +64,35 @@ public AnnotatedViewAccessChecker(
...
+                    .orElse(true);
+            if (noParents
+                    && registry.hasLayout(context.getLocation().getPath())) {
+                Class<?> layout = registry
+                        .getLayout(context.getLocation().getPath());
+
+                boolean hasAccess = accessAnnotationChecker.hasAccess(layout,
+                        context.getPrincipal(), context::hasRole);
+                if (!hasAccess) {
+                    LOGGER.info(
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/auth/AnnotatedViewAccessChecker.java]: done

File: flow-server/src/main/java/com/vaadin/flow/server/auth/AnnotatedViewAccessChecker.java [hunk 2]
```diff
@@ -74,6 +107,18 @@ public AccessCheckResult check(NavigationContext context) {
             denyReason = "Consider adding one of the following annotations "
                     + "to make the view accessible: @AnonymousAllowed, "
                     + "@PermitAll, @RolesAllowed.";
+            if (targetView.isAnnotationPresent(Layout.class)
+                    && context.getRouter().getRegistry()
+                            .getTargetUrl(
+                                    (Class<? extends Component>) targetView)
+                            .isEmpty()) {
+                LOGGER.info(
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/auth/AnnotatedViewAccessChecker.java]: done

File: flow-server/src/test/java/com/vaadin/flow/router/internal/RouteUtilTest.java [hunk 3]
```diff
@@ -316,11 +319,40 @@ public VaadinContext getContext() {
...
+        ApplicationRouteRegistry registry = ApplicationRouteRegistry
+                .getInstance(service.getContext());
+        registry.setLayout(AutoLayout.class);
+
+        List<Class<? extends RouterLayout>> parentLayouts = RouteUtil
+                .getParentLayouts(registry, AutoLayoutView.class, "auto");
+
+        Assert.assertEquals(
+                "Route with no layout should not get automatic layout", 0,
+                parentLayouts.size());
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/test/java/com/vaadin/flow/router/internal/RouteUtilTest.java]: done

**Comment by caalador** [comment]
The `@Route` + `@Layout` on the same class would probably work if you add the `autoLayout=false`. We should perhaps add a check to not layout the route with itself.

**Review by mshabarov — APPROVED** [review]
LGTM. Couldn't find anything suspicious or unexpected.
I managed to test with URL-based access control, works as expected and nothing odd there. URL-based control is basically made by security framework, Spring Security in my case and isn't directly related to this.

About @Layout + @Route, I propose to restrict this use case for now, until this would requested in the future.

**Pull request merged**